scripts/5.hetnet-pathways.py

```diff
@@ -0,0 +1,123 @@
+
+# In[2]:
+
+query = '''
+MATCH path = (gene:Gene)
+  -[:PARTICIPATES_GpPW]-(pathway:Pathway)
+RETURN gene.name as gene_name,
+  gene.identifier as gene_id, 
+  pathway.name as pathway_name,
+  pathway.identifier as pathway_id
+'''
+
+driver = GraphDatabase.driver("bolt://neo4j.het.io")
+hetnet_results = pd.DataFrame()
+with driver.session() as session:
+    result = session.run(query)
+    hetnet_results = ( pd.DataFrame((x.values() for x in result), columns=result.keys())
+                       .sort_values(by='pathway_id') )
+
+
+# In[3]:
+
+print('There were %d gene-pathway interactions in the hetnet query.' % len(hetnet_results))
+
+
+# In[4]:
+
+hetnet_results.head()
+
+
+# In[5]:
+
+path = os.path.join('data', 'mutation-matrix.tsv.bz2')
+mutation_df = pd.read_table(path, index_col=0)
+mutation_df.head()
+
+
+# Note that columns of <code>mutation_df</code> are strings. After a quick search I couldn't find out how to fix this, so be warned that code below works around this.
+
+# In[6]:
+
+type(mutation_df.columns[0])
+
+
+# Display some information about the two datasets...
+
+# In[7]:
+
+cognoma_genes = set([int(gene_id) for gene_id in mutation_df.columns])
+print('There were %d genes found in the Cognoma dataset.' % len(cognoma_genes))
+
+hetnet_genes = set(hetnet_results['gene_id'])
+print('There were %d genes found in the hetnet pathway query.' % len(hetnet_genes))
+
+genes_in_both = cognoma_genes.intersection(hetnet_genes)
+print('There were %d genes found in both the Cognoma dataset and the hetnet query.' % len(genes_in_both))
+
+pathways = set(hetnet_results['pathway_id'])
+print('There were %d pathways found in the hetnet pathway query.' % len(pathways))
+
+number_missing = len(cognoma_genes-hetnet_genes)
+print('There were %d genes in the Cognoma dataset that were not found in the hetnet pathway query.' % number_missing)
+
+number_missing = len(hetnet_genes-cognoma_genes)
+print('There were %d genes in the hetnet pathway query that were not found in the Cognoma dataset.' % number_missing)
+
+
+# Initialize a samples by pathways data frame...
+
+# In[8]:
+
+number_of_samples = len(mutation_df)
+number_of_pathways = len(pathways) 
+sample_pathway_df = pd.DataFrame(np.zeros((number_of_samples, number_of_pathways), dtype=np.int),
+                                 index=mutation_df.index,
+                                 columns=pathways)
+
+
+# Now populate this data frame. This is a slow Python loop, hence the progress bar. It takes a few minutes on my laptop. The idea is to loop over all gene-pathway interactions in the hetnet query. If the gene is in the Cognoma dataset, we grab the pathway id in that gene-pathway interaction. We look at Cognoma samples where that gene is labeled 1, i.e., at Cognoma samples that have a mutation in that gene, and grab the corresponding indices. Then, in the pathway matrix all samples get the associated pathway tagged as a 1, since they have a mutated gene that participates in that pathway.
```

Comment:
I would consider the following strategy:

1. melt mutation_df so you get at `sample_id, entrez_gene_id, mutation` dataframe.
2. filter that for `mutation == 1`
3. merge with `pathway_df`
4. pivot_table to create a sample by pathway matrix. See [this example](https://github.com/cognoma/cancer-data/blob/54140cf6addc48260c9723213c40b628d7c861da/scripts/2.TCGA-process.py#L157-L161) for this step.

Hopefully that's a little bit faster and more readable. You won't be able to use the progress bar anymore.